File: OsmAnd/src/net/osmand/plus/mapcontextmenu/builders/MergeLocalizedTagsAlgorithm.java
package net.osmand.plus.mapcontextmenu.builders;

import androidx.annotation.NonNull;

import net.osmand.osm.AbstractPoiType;
import net.osmand.osm.MapPoiTypes;
import net.osmand.plus.OsmandApplication;

import java.util.ArrayList;
import java.util.Arrays;
import java.util.HashMap;
import java.util.List;
import java.util.Map;

public class MergeLocalizedTagsAlgorithm {

	private static final List<String> NAME_TAG_PREFIXES = Arrays.asList(
			"name", "int_name", "nat_name", "reg_name", "loc_name",
			"old_name", "alt_name", "short_name", "official_name", "lock_name"
	);

	private final OsmandApplication app;

	private MergeLocalizedTagsAlgorithm(@NonNull OsmandApplication app) {
		this.app = app;
	}

	@NonNull
	public static Map<String, Object> execute(@NonNull OsmandApplication app,
	                                          @NonNull Map<String, String> originalDict) {
		MergeLocalizedTagsAlgorithm instance = new MergeLocalizedTagsAlgorithm(app);
		return instance.executeImpl(new HashMap<>(originalDict));
	}

	@NonNull
	private Map<String, Object> executeImpl(@NonNull Map<String, Object> originalDict) {
		Map<String, Object> resultDict = new HashMap<>();
		Map<String, Map<String, Object>> localizationsDict = new HashMap<>();

		for (String key : originalDict.keySet()) {
			processAdditionalTypeWithKey(key, convertKey(key), originalDict, localizationsDict, resultDict);
		}

		List<String> keysToUpdate = findKeysToUpdate(localizationsDict);
		for (String baseKey : keysToUpdate) {
			Map<String, Object> localizations = localizationsDict.get(baseKey);
			if (localizations != null) {
				localizations.put(baseKey, originalDict.get(baseKey));
			}
		}
		Map<String, Object> finalDict = finalizeLocalizationDict(localizationsDict);
		addRemainingEntriesFrom(resultDict, finalDict);
		return finalDict;
	}

	private void processNameTagWithKey(@NonNull String key,
	                                   @NonNull String convertedKey,
	                                   @NonNull Map<String, Object> originalDict,
	                                   @NonNull Map<String, Map<String, Object>> localizationsDict) {
		if (key.contains(":")) {
			String[] components = convertedKey.split(":");
			if (components.length == 2) {
				String baseKey = components[0];
				String localeKey = baseKey + ":" + components[1];

				Map<String, Object> nameDict = dictionaryForKey("name", localizationsDict);
				nameDict.put(localeKey, originalDict.get(convertedKey));
			}
		} else {
			Map<String, Object> nameDict = dictionaryForKey("name", localizationsDict);
			nameDict.put(convertedKey, originalDict.get(key));
		}
	}

	private void processAdditionalTypeWithKey(@NonNull String key,
	                                          @NonNull String convertedKey,
	                                          @NonNull Map<String, Object> originalDict,
	                                          @NonNull Map<String, Map<String, Object>> localizationsDict,
	                                          @NonNull Map<String, Object> resultDict) {
		MapPoiTypes poiTypes = app.getPoiTypes();
		AbstractPoiType poiType = poiTypes.getAnyPoiAdditionalTypeByKey(convertedKey);
		if (poiType != null && poiType.getLang() != null && key.contains(":")) {
			String[] components = key.split(":");
			if (components.length == 2) {
				String baseKey = components[0];
				String localeKey = baseKey + ":" + components[1];

				Map<String, Object> baseDict = dictionaryForKey(baseKey, localizationsDict);
				baseDict.put(localeKey, originalDict.get(key));
			}
		} else {
			resultDict.put(key, originalDict.get(key));
		}
	}

	@NonNull
	private Map<String, Object> dictionaryForKey(@NonNull String key, @NonNull Map<String, Map<String, Object>> dict) {
		return dict.computeIfAbsent(key, k -> new HashMap<>());
	}

	@NonNull
	private List<String> findKeysToUpdate(@NonNull Map<String, Map<String, Object>> localizationsDict) {
		List<String> keysToUpdate = new ArrayList<>();
		for (String baseKey : localizationsDict.keySet()) {
			Map<String, Object> localizations = localizationsDict.get(baseKey);
			if (localizations != null && !localizations.containsKey(baseKey)) {
				keysToUpdate.add(baseKey);
			}
		}
		return keysToUpdate;
	}

	@NonNull
	private Map<String, Object> finalizeLocalizationDict(@NonNull Map<String, Map<String, Object>> localizationsDict) {
		Map<String, Object> finalDict = new HashMap<>();

		for (String baseKey : localizationsDict.keySet()) {
			Map<String, Object> entryDict = new HashMap<>();
			Map<String, Object> localizations = localizationsDict.get(baseKey);
			entryDict.put("localizations", localizations);
			finalDict.put(baseKey, entryDict);
		}
		return finalDict;
	}

	private void addRemainingEntriesFrom(@NonNull Map<String, Object> resultDict,
	                                     @NonNull Map<String, Object> finalDict) {
		for (String key : resultDict.keySet()) {
			finalDict.putIfAbsent(key, resultDict.get(key));
		}
	}

	public boolean isNameTag(String tag) {
		for (String prefix : NAME_TAG_PREFIXES) {
			if (tag.startsWith(prefix)) {
				return true;
			}
		}
		return false;
	}

	private String convertKey(String key) {
		return key.replace("_-_", ":");
	}
}

File: OsmAnd/src/net/osmand/plus/mapcontextmenu/builders/MergeLocalizedTagsAlgorithm.kt
package net.osmand.plus.mapcontextmenu.builders

import net.osmand.data.Amenity.DESCRIPTION
import net.osmand.data.Amenity.LANG_YES
import net.osmand.data.Amenity.NAME
import net.osmand.osm.AbstractPoiType
import net.osmand.osm.MapPoiTypes
import net.osmand.plus.OsmandApplication
import net.osmand.shared.gpx.GpxFile.Companion.XML_COLON

private val NAME_TAG_PREFIXES = listOf(
	"name", "int_name", "nat_name", "reg_name", "loc_name",
	"old_name", "alt_name", "short_name", "official_name", "lock_name"
)

class MergeLocalizedTagsAlgorithm private constructor(private val app: OsmandApplication) {

	companion object {
		fun execute(
			app: OsmandApplication,
			originalDict: Map<String, String>
		): Map<String, Any> {
			val instance = MergeLocalizedTagsAlgorithm(app)
			return instance.executeImpl(HashMap(originalDict))
		}
	}

	private fun executeImpl(originalDict: Map<String, Any>): Map<String, Any> {
		val resultDict = mutableMapOf<String, Any>()
		val langDict = applyLanguageOverride(originalDict)
		val localizationsDict = mutableMapOf<String, MutableMap<String, Any>>()

		for (key in langDict.keys) {
			processAdditionalTypeWithKey(
				key, convertKey(key), langDict, localizationsDict, resultDict
			)
		}

		val keysToUpdate = findKeysToUpdate(localizationsDict)
		for (baseKey in keysToUpdate) {
			val localizations = localizationsDict[baseKey]
			localizations?.put(baseKey, langDict[baseKey] ?: continue)
		}

		val finalDict = finalizeLocalizationDict(localizationsDict)
		addRemainingEntriesFrom(resultDict, finalDict)
		return finalDict
	}

	private fun applyLanguageOverride(originalDict: Map<String, Any>): MutableMap<String, Any> {
		val processedDict = originalDict.toMutableMap()

		val langKeys = processedDict.keys.filter {
			it.startsWith("$LANG_YES:") && processedDict[it] == "yes"
		}
		if (langKeys.size == 1) {
			val langKey = langKeys.first()
			val langCode = langKey.substringAfter("$LANG_YES:")

			renameTagInMap(processedDict, NAME, "$NAME:$langCode")
			renameTagInMap(processedDict, DESCRIPTION, "$DESCRIPTION:$langCode")
		}
		return processedDict
	}

	private fun renameTagInMap(dict: MutableMap<String, Any>, oldKey: String, newKey: String) {
		val value = dict.remove(oldKey)
		if (value != null) {
			dict[newKey] = value
		}
	}

	private fun processNameTagWithKey(
		key: String,
		convertedKey: String,
		originalDict: Map<String, Any>,
		localizationsDict: MutableMap<String, MutableMap<String, Any>>
	) {
		if (key.contains(":")) {
			val components = convertedKey.split(":")
			if (components.size == 2) {
				val baseKey = components[0]
				val localeKey = "$baseKey:${components[1]}"

				val nameDict = dictionaryForKey("name", localizationsDict)
				nameDict[localeKey] = originalDict[convertedKey] ?: return
			}
		} else {
			val nameDict = dictionaryForKey("name", localizationsDict)
			nameDict[convertedKey] = originalDict[key] ?: return
		}
	}

	private fun processAdditionalTypeWithKey(
		key: String,
		convertedKey: String,
		originalDict: Map<String, Any>,
		localizationsDict: MutableMap<String, MutableMap<String, Any>>,
		resultDict: MutableMap<String, Any>
	) {
		val poiTypes: MapPoiTypes = app.poiTypes
		val poiType: AbstractPoiType? = poiTypes.getAnyPoiAdditionalTypeByKey(convertedKey)
		if (poiType?.lang != null && key.contains(":")) {
			val components = key.split(":")
			if (components.size == 2) {
				val baseKey = components[0]
				val localeKey = "$baseKey:${components[1]}"

				val baseDict = dictionaryForKey(baseKey, localizationsDict)
				baseDict[localeKey] = originalDict[key] ?: return
			}
		} else {
			resultDict[key] = originalDict[key] ?: return
		}
	}

	private fun dictionaryForKey(
		key: String,
		dict: MutableMap<String, MutableMap<String, Any>>
	): MutableMap<String, Any> {
		return dict.getOrPut(key) { mutableMapOf() }
	}

	private fun findKeysToUpdate(localizationsDict: Map<String, MutableMap<String, Any>>): List<String> {
		val keysToUpdate = mutableListOf<String>()
		for (baseKey in localizationsDict.keys) {
			val localizations = localizationsDict[baseKey]
			if (localizations != null && !localizations.containsKey(baseKey)) {
				keysToUpdate.add(baseKey)
			}
		}
		return keysToUpdate
	}

	private fun finalizeLocalizationDict(localizationsDict: Map<String, MutableMap<String, Any>>): MutableMap<String, Any> {
		val finalDict = mutableMapOf<String, Any>()

		for (baseKey in localizationsDict.keys) {
			val entryDict = mutableMapOf<String, Any>()
			val localizations = localizationsDict[baseKey]
			entryDict["localizations"] = localizations ?: continue
			finalDict[baseKey] = entryDict
		}
		return finalDict
	}

	private fun addRemainingEntriesFrom(
		resultDict: Map<String, Any>,
		finalDict: MutableMap<String, Any>
	) {
		for (key in resultDict.keys) {
			finalDict.putIfAbsent(key, resultDict[key] ?: continue)
		}
	}

	fun isNameTag(tag: String): Boolean {
		for (prefix in NAME_TAG_PREFIXES) {
			if (tag.startsWith(prefix)) {
				return true
			}
		}
		return false
	}

	private fun convertKey(key: String): String {
		return key.replace(XML_COLON, ":")
	}
}
